File: src/Traap/DataTypes.hs
-- | Copyright (c) Gary Allan Howard aka Traap.
-- License BSD-3-Clause
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}

module Traap.DataTypes
  (-- Symbolic Links
   Symlinks(..)
  ,Symlink(..)
  ,toSymlinks
  ,toSymlink
  -- Git Repositories
  ,Repos(..)
  ,Repo(..)
  ,Path(..)
  ,toRepos
  ,toRepo
  ,toPath
  -- Programs to install.
  ,Installations(..)
  ,Os(..)
  ,Command(..)
  ,toInstallations
  ,toOs
  ,toCommand
  ) where

import           Data.Yaml
import qualified Data.Text as T
import           GHC.Generics

-- -----------------------------------------------------------------------------
-- Actions bootstrap can perform.
data Action = Create | Delete | Clone | Install

-- -----------------------------------------------------------------------------
-- | The Symlinks type defines the target of a symlink operation and identifies
-- the target type as a file or as a directory.
data Symlinks = SYMLINKS
  { symlinks :: [Symlink]
  } deriving (Show, Generic, FromJSON)

data Symlink = SYMLINK
  { file :: T.Text
  , link :: T.Text
  , directory :: Bool
  } deriving (Show, Generic, FromJSON)

toSymlinks :: Symlinks -> Action -> [T.Text]
toSymlinks s a = case a of
  Create -> map toSymlink (symlinks s)
  Delete -> map toSymlink' (symlinks s)

toSymlink :: Symlink -> T.Text
toSymlink s = mconcat ["lns -s ", file s, " ", link s]

toSymlink' :: Symlink -> T.Text
toSymlink' s = mconcat ["rm -vrf ", link s]

-- -----------------------------------------------------------------------------
-- | The Repos type defines a Repo and whether or not the Repo is cloned to .
-- (here) or a new directory.
data Repos = REPOS
  { repos :: [Repo]
  } deriving (Show, Generic, FromJSON)

data Repo = REPO
  { url :: T.Text
  , paths ::  [Path]
  } deriving (Show, Generic, FromJSON)

data Path = PATH
  { source :: T.Text
  , target :: T.Text
  } deriving (Show, Generic, FromJSON)

toRepos :: Repos -> Action -> [T.Text]
toRepos r a = concatMap (toRepo a) (repos r)

toRepo :: Repo -> Action -> [T.Text]
toRepo REPO {..} a = map (toPath url a) paths

toPath :: T.Text -> Path -> Action -> T.Text
toPath u p a = case a of 
  Clone -> mconcat ["git clone ", u, "/", source p, target p]
  Delete -> mconcat [target p]

-- -----------------------------------------------------------------------------
-- | The Installations type defines a program to run and the argument is is passed.
data Installations = INSTALLATIONS
  { installations :: [Os]
  } deriving (Show, Generic, FromJSON)

data Os = OS
  { name :: T.Text
  , command :: [Command]
  } deriving (Show, Generic, FromJSON)

data Command = COMMAND
  { sudo :: Bool
  , program :: T.Text
  , argument :: T.Text
  } deriving (Show, Generic, FromJSON)

toInstallations :: Installations -> Action -> [T.Text]
toInstallations = concatMap toOs . installations

toOs :: Os -> Action -> [T.Text]
toOs = map toCommand . command

-- TODO: Needs operating system and sudo awareness.
toCommand :: Command -> Action -> T.Text
toCommand c a = case a of Install -> mconcat [program c, argument c]

File: ruby/DataTypes.rb
#!/usr/local/bin/ruby
# Copyright (c) Gary Allan Howard aka Traap.
# License BSD-3-Clause
# ------------------------------------------------------------------------------

require 'open3'
require 'yaml'

# ------------------------------------------------------------------------------

class ShellError < StandardError; end

# ------------------------------------------------------------------------------
class Command
  def initialize(data, options)
    @data = data
    @options = options 
    @command = nil
  end
  
  def install_artifact; end

  def remove_artifact; end

  def do_command
    echo_command if @options.verbose || @options.dryrun
    run_command  if !@options.dryrun
  end

  def echo_command
    puts @command
  end

  def run_command
    begin
      status = system(@command)
    rescue ShellError
      abort "System command failed: #{status}"
    end
  end

end # End Command

# ------------------------------------------------------------------------------
class SymLink < Command
  def initialize(data, options)
    super(data, options)
  end

  def remove_artifact 
    puts "Deleting symbolic links"
    @data.each do |n|
      n['symlink'].each do |s|
       @command = "rm -frv #{s['link']}"
       do_command
      end
    end
    puts ""
  end

  def install_artifact 
    puts "Making symbolic links"
    @data.each do |n|
      n['symlink'].each do |s|
        @command = "ln -s #{s['file']} #{s['link']}"
        do_command
      end
    end
    puts ""
  end
end # End SymLink

# ------------------------------------------------------------------------------
class Repo < Command
  def initialize(data, options)
    super(data, options)
  end

  def install_artifact 
    puts "Cloning repositories"
    @data.each do |n|
      n['paths'].each do |p|
        @command = "git clone #{n['url']}/#{p['source']} #{p['target']}"
        do_command
      end
    end
    puts ""
  end
end # End Repo

# ------------------------------------------------------------------------------
class Install < Command
  def initialize(data, options)
    super(data, options)
  end
  
  def install_artifact
    puts "Installing programs"
    @data.each do |n|
      n['command'].each do |c|
        if c['sudo'] then
          @command = "sudo #{c['program']} #{c['argument']}"
        else
          @command ="#{c['program']} #{c['argument']}"
        end
        do_command
      end
    end
    puts ""
  end
end # End Install
# ------------------------------------------------------------------------------
